### PR TITLE
add cells, rows, columns functions

### PR DESCRIPTION
**Pull request by lesliekim**

### CODE REVIEW

File: demos/grid-view/index.js [hunk 1]
```diff
@@ -10,6 +10,7 @@ import {
   cells,
 } from '../../js/vnext/projection';
 import './index.less';
+import * as tmpl from './bodyTemplate.jade';
```
Comment (lyweiwei) [inline demos/grid-view/index.js]:
``` js
import bodyTemplate from './body-template.jade';
```
1. File name should use kabab-case
2. Use don't use short names for globals

File: js/vnext/projection/cells.js [hunk 2]
```diff
@@ -12,10 +12,19 @@ function translateRow(columnGroup, row) {
     };
   }
   if (_.has(row, 'item')) {
-    return {
+    const obj =  {
       classes: row.classes,
-      cells: _.map(columnGroup.leafColumns, col => (row.item[col.name] || {})),
+      cells: _.map(columnGroup.leafColumns, col => {
+        const cell = row.item[col.name] || {};
+        const tpl = col.template || col.headTemplate || col.footTemplate;
```
Comment (lyweiwei) [inline js/vnext/projection/cells.js]: This is incorrect, you should pick the right template for given row group.

File: js/vnext/projection/cells.js [hunk 3]
```diff
@@ -12,10 +12,19 @@ function translateRow(columnGroup, row) {
...
+        if (tpl) {
+          cell.html =  _.isFunction(tpl) ? tpl(row.item) : tpl;
+        }
+        return cell;
```
Comment (lyweiwei) [inline js/vnext/projection/cells.js]: You can not return the `row.item[col.name]` directly, it must be a separated copy

File: js/vnext/projection/column-group.js [hunk 4]
```diff
@@ -8,13 +8,15 @@ class ColumnGroup {
...
       parent = null,
       columns = [],
       html = name,
       height = 1,
     }) => {
-      const col = { name, width, parent, html, height };
-
+      const col = { name, width, template, parent, html, height };
+      
+      col.template = template ? template : (parent? parent.template : null);
```
Comment (lyweiwei) [inline js/vnext/projection/column-group.js]: don't inherit the template, BTW, the key should be 'headTemplate'

File: js/vnext/projection/column-group.js [hunk 5]
```diff
@@ -8,13 +8,15 @@ class ColumnGroup {
...
       width,
+      template,
       parent = null,
       columns = [],
       html = name,
       height = 1,
     }) => {
-      const col = { name, width, parent, html, height };
-
+      const col = { name, width, template, parent, html, height };
```
Comment (lyweiwei) [inline js/vnext/projection/column-group.js]: All the configurations need to be included:
- headTemplate
- headHtml
- bodyTemplate
- bodyHtml
- footTemplate
- footHtml

File: js/vnext/projection/jsdata.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import _ from 'underscore';
+import Promise from 'bluebird';
+
+export function jsdataProj (p$state, {
```
Comment (lyweiwei) [inline js/vnext/projection/jsdata.js]: `jsdataProj` -> `jsdata`

Comment (lyweiwei) [inline js/vnext/projection/jsdata.js]: `p$state` -> `state`, as it's not a promise

File: js/vnext/projection/jsdata.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+	query,
+	entity,
+	options,
+	skip,
+	take,
+	filter,
+	orderby = [],
```
Comment (lyweiwei) [inline js/vnext/projection/jsdata.js]: Fix indentation.

File: js/vnext/projection/jsdata.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+	if (orderby && orderby.length) {
+		op.orderby = _.reduce(orderby, (memo, obj) => {
+			_.each(obj, (key, value) => {memo.push([key, value > 0 ? 'ASC' : 'DESC']);});
+			return memo;
+		}, []);
+	}
+
+	return entity.fincAll(op, _.defaults(options, { all: true }))
+	.then(_.property('value'));
```
Comment (lyweiwei) [inline js/vnext/projection/jsdata.js]: Fix indentation

File: js/vnext/projection/rows.js [hunk 9]
```diff
@@ -11,5 +12,17 @@ export const rows = (state, {
 
   return { headRows, bodyRows, footRows, columns: state.columns };
 };
+*/
+export function rows(state, {
+  headRows = ['column-header-rows'],
+  footRows = [],
+} = {}) {
+  const columns = state.columns;
+  const bodyRows = _.map(_.omit(state, 'columns'), item => ({
```
Comment (lyweiwei) [inline js/vnext/projection/rows.js]: you don't need to call omit, I have a change to fix this, let me push first.

File: demos/grid-view/index.js (hunk 1 above)
Comment (lesliekim) [inline demos/grid-view/index.js]: It's fixed~

File: js/factory/renderers-plugin.js [hunk 10]
```diff
@@ -2,17 +2,17 @@ import layout from '../layout/index';
 
 export default definePlugin => definePlugin('renderers', [
   'config',
-], function (config) {
-  const renderers = [];
+  ], function (config) {
```
Comment (lyweiwei) [inline js/factory/renderers-plugin.js:5]: not intended to change this file?

File: js/vnext/projection/cells.js [hunk 11]
```diff
@@ -12,10 +12,28 @@ function translateRow(columnGroup, row) {
...
   }
   if (_.has(row, 'item')) {
-    return {
+    const obj =  {
       classes: row.classes,
-      cells: _.map(columnGroup.leafColumns, col => (row.item[col.name] || {})),
+      cells: _.map(columnGroup.leafColumns, col => {
+        const cell = row.item[col.name] || {};
+        const classes = col.classes;
+        const template = col[group + 'Template'];
```
Comment (lyweiwei) [inline js/vnext/projection/cells.js:20]: give it a default value, _.identity

Comment (lyweiwei) [inline js/vnext/projection/cells.js:20]: BTW, the template should always be function, don't need to check it.

``` js
const html = _.result(cell, 'html', template(group === 'head'? col, row.item));
```

File: js/vnext/projection/column-group.js [hunk 12]
```diff
@@ -5,25 +5,19 @@ class ColumnGroup {
...
-      const col = { name, width, parent, html, height };
-
+    const buildColumn = col => {
+      const { parent, columns, height } = col;
+      
+      col.height  = _.isNumber(height) ? height : 1;
       col.rowIndex = parent ? parent.rowIndex + parent.height : 0;
       col.columns = _.map(columns, c => buildColumn(_.extend({ parent: col }, c)));
-      col.treeHeight = height;
+      col.treeHeight = _.isNumber(height) ? height : 1;
```
Comment (lyweiwei) [inline js/vnext/projection/column-group.js:14]: col.treeHeight = col.height

File: js/vnext/projection/column-group.js [hunk 13]
```diff
@@ -5,25 +5,19 @@ class ColumnGroup {
...
       col.treeWidth = 1;
       if (!_.isEmpty(col.columns)) {
         col.treeHeight += _.chain(col.columns)
-          .map(_.property('treeHeight')).max().value();
+        .map(_.property('treeHeight')).max().value();
```
Comment (lyweiwei) [inline js/vnext/projection/column-group.js:18]: indent needed

File: js/vnext/projection/column-group.js [hunk 14]
```diff
@@ -37,12 +31,13 @@ class ColumnGroup {
       if (col.parent) {
         const colspan = col.treeWidth;
         const rowspan = _.isEmpty(col.columns) ? this.root.treeHeight - col.rowIndex : col.height;
-        const html = col.html;
-
+        const html = col.html || col.name;
+        console.log(this.headerRows);
+        console.log(col.rowIndex);
```
Comment (lyweiwei) [inline js/vnext/projection/column-group.js:36]: you're going to remove this :)

File: js/vnext/projection/jsdata.js [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+  if (orderby && orderby.length) {
+    op.orderby = _.reduce(orderby, (memo, obj) => {
+      _.each(obj, (key, value) => {memo.push([key, value > 0 ? 'ASC' : 'DESC']);});
+      return memo;
+    }, []);
+  }
+
+  return entity.fincAll(op, _.defaults(options, { all: true }))
+  .then(_.property('value'));
```
Comment (lyweiwei) [inline js/vnext/projection/jsdata.js:40]: indent by 2

File: js/vnext/projection/rows.js [hunk 16]
```diff
@@ -4,11 +4,11 @@ export function rows(state, {
...
   footRows = [],
 } = {}) {
+  const columns = state.columns;
   const bodyRows = _.map(state.items, item => ({
     classes: ['body-row'],
-    item: _.mapObject(item, value => ({ html: value })),
-  }));
-
-  return { headRows, bodyRows, footRows, columns: state.columns };
+    item:  item, //_.mapObject(item, value => ({ html: value })),
```
Comment (lyweiwei) [inline js/vnext/projection/rows.js:10]: This can simply be

``` javascript
{
  classes: ['body-row'],
  item,
}
```

**Comment by lyweiwei** [comment]
@lesliekim, There's no critical feedbacks, I'll merge the code in. You can fix the comments in a following change.
BTW, you can ignore the comments in the memory/odata/jsdata projections as I'll merge them into dataSource projection.

File: js/vnext/projection/jsdata.js (hunk 15 above)
Comment (lesliekim) [inline js/vnext/projection/jsdata.js:40]: sublime auto reindent does not work on this file~ So I fix the indentation by hand T_T

File: js/vnext/projection/column-group.js (hunk 13 above)
Comment (lesliekim) [inline js/vnext/projection/column-group.js:18]: Er.....sublime reindent removes these indent

**Pull request merged**